### ISSUE TITLE
DAG Task with Skipped Dependency Still Executes

### ISSUE THREAD

**Issue by terry-sm** [argoproj/argo#1386]
**Is this a BUG REPORT or FEATURE REQUEST?**: 
Bug Report

**What happened**: 
A DAG workflow task with dependencies executed when its dependency was skipped. The dependency task was skipped due to the `when` condition. For example:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: dag-skipped-dep-
spec:
  entrypoint: workflow
  templates:
    - name: workflow
      dag:
        tasks:
          - name: A
            template: whalesay
            arguments:
              parameters:
                - name: message
                  value: "From A"
          - name: B
            template: whalesay
            dependencies: [A]
            when: "'1' == '2'"
            arguments:
              parameters:
                - name: message
                  value: "From B"
          - name: C
            template: whalesay
            dependencies: [B]
            arguments:
              parameters:
                - name: message
                  value: "From C"
    - name: whalesay
      inputs:
        parameters:
          - name: message
      container:
        image: docker/whalesay:latest
        command: [cowsay]
        args: ["{{inputs.parameters.message}}"]

```

When I run this job A runs, B is skipped, and C runs despite the dependency.

This happens because Argo workflow controller considers skipped as successful. 

**What you expected to happen**:
I expected the task that had a skipped dependency to also be skipped. If Task A was skipped and Task A is a dependency of Task B, I would expect Task B to also be skipped. 

**How to reproduce it (as minimally and precisely as possible)**:
See example above. 

**Anything else we need to know?**:

This is the block in question for 2.2.1 - https://github.com/argoproj/argo/blob/0a928e93dac6d8522682931a0a68c52add310cdb/workflow/controller/dag.go#L230-L232

which uses this method: https://github.com/argoproj/argo/blob/0a928e93dac6d8522682931a0a68c52add310cdb/pkg/apis/workflow/v1alpha1/types.go#L550-L552

In 2.3.0, the code changed slightly, but the condition is still there: `depNode.Successful()` so it will still occur in 2.3.0: https://github.com/argoproj/argo/blob/88fcc70dcf6e60697e6716edc7464a403c49b27e/workflow/controller/dag.go#L251-L253

**Environment**:
- Argo version: 
```
argo: v2.2.1
  BuildDate: 2018-10-11T16:25:59Z
  GitCommit: 3b52b26190163d1f72f3aef1a39f9f291378dafb
  GitTreeState: clean
  GitTag: v2.2.1
  GoVersion: go1.10.3
  Compiler: gc
  Platform: darwin/amd64
```

- Kubernetes version :
```
clientVersion:
  buildDate: "2019-03-01T23:34:27Z"
  compiler: gc
  gitCommit: c27b913fddd1a6c480c229191a087698aa92f0b1
  gitTreeState: clean
  gitVersion: v1.13.4
  goVersion: go1.12
  major: "1"
  minor: "13"
  platform: darwin/amd64
serverVersion:
  buildDate: "2019-02-28T18:40:05Z"
  compiler: gc
  gitCommit: 4e209c9383fa00631d124c8adcc011d617339b3c
  gitTreeState: clean
  gitVersion: v1.11.8
  goVersion: go1.10.8
  major: "1"
  minor: "11"
  platform: linux/amd64
```

**Other debugging information (if applicable)**:
- workflow result:
```
Name:                dag-skipped-dep-b5zt7
Namespace:           yolo
ServiceAccount:      default
Status:              Succeeded
Created:             Fri May 31 13:03:43 -0500 (5 minutes ago)
Started:             Fri May 31 13:03:43 -0500 (5 minutes ago)
Finished:            Fri May 31 13:04:15 -0500 (5 minutes ago)
Duration:            32 seconds

STEP                      PODNAME                           DURATION  MESSAGE
 ✔ dag-skipped-dep-b5zt7
 ├-✔ A                    dag-skipped-dep-b5zt7-2660881670  19s
 ├-○ B                                                                when ''1' == '2'' evaluated false
 └-✔ C                    dag-skipped-dep-b5zt7-2627326432  10s
```

**Comment by CermakM** [comment]
+1, this is quite unpleasant

**Comment by ptitpoulpe** [comment]
+1

**Comment by simster7** [comment]
#2673 Should provide functionality to fix this

**Comment by alexec** [comment]
@simster7 to double check and maybe close.

**Comment by simster7** [comment]
This can be solved using the new depends logic in #2673
